jabgui/src/main/java/org/jabref/gui/externalfiles/AutoRenameFileOnEntryChange.java

```diff
@@ -0,0 +1,66 @@
+package org.jabref.gui.externalfiles;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.jabref.gui.preferences.GuiPreferences;
+import org.jabref.logic.FilePreferences;
+import org.jabref.logic.citationkeypattern.CitationKeyGenerator;
+import org.jabref.logic.cleanup.RenamePdfCleanup;
+import org.jabref.model.database.BibDatabaseContext;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.event.FieldChangedEvent;
+
+import com.google.common.eventbus.Subscribe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.jabref.logic.citationkeypattern.BracketedPattern.expandBrackets;
+
+public class AutoRenameFileOnEntryChange {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AutoRenameFileOnEntryChange.class);
+
+    private final GuiPreferences preferences;
+    private final BibDatabaseContext bibDatabaseContext;
+    private final RenamePdfCleanup renamePdfCleanup;
+
+    public AutoRenameFileOnEntryChange(BibDatabaseContext bibDatabaseContext, GuiPreferences preferences) {
+        this.bibDatabaseContext = bibDatabaseContext;
+        this.preferences = preferences;
+        bibDatabaseContext.getDatabase().registerListener(this);
+        renamePdfCleanup = new RenamePdfCleanup(false, () -> bibDatabaseContext, preferences.getFilePreferences());
+    }
+
+    @Subscribe
+    public void listen(FieldChangedEvent event) {
+        FilePreferences filePreferences = preferences.getFilePreferences();
+
+        if (!filePreferences.shouldAutoRenameFilesOnChange()
+                || filePreferences.getFileNamePattern().isEmpty()
+                || filePreferences.getFileNamePattern() == null
+                || !relatesToFilePattern(filePreferences.getFileNamePattern(), event)) {
+            return;
+        }
+
+        BibEntry entry = event.getBibEntry();
+        if (entry.getFiles().isEmpty()) {
+            return;
+        }
+        new CitationKeyGenerator(bibDatabaseContext, preferences.getCitationKeyPatternPreferences()).generateAndSetKey(entry);
+        renamePdfCleanup.cleanup(entry);
+
+        LOGGER.info("Field changed for entry {}: {}", entry.getCitationKey().orElse("defaultCitationKey"), event.getField().getName());
+    }
+
+    private boolean relatesToFilePattern(String fileNamePattern, FieldChangedEvent event) {
+        Set<String> extractedFields = new HashSet<>();
+
+        expandBrackets(fileNamePattern, bracketContent -> {
+            extractedFields.add(bracketContent);
```

Comment:
This is too straight forward. See for instance https://docs.jabref.org/setup/citationkeypatterns#modifiers

Maybe, call `FileUtil.createFileNameFromPattern(databaseContext.getDatabase(), entry, filePreferences.getFileNamePattern()).trim()` to check if there will be a move action

Maybe, you can also leave this out, because `if (suggestedFileName.equals(currentFileName)) {` already checks for equality. (One could rewrite `org.jabref.logic.externalfiles.LinkedFileHandler#renameToSuggestedName` though --> first check the file names to avoid the lookup)